README
NAME
    Catalyst::ActionRole::PseudoCache - Super simple caching for Catalyst
    actions

VERSION
    version 1.000002

SYNOPSIS
     package MyApp::Controller::Root;

     use Moose;
     BEGIN { extends 'Catalyst::Controller::ActionRole' };

     # used with Catalyst::Plugin::Cache
     sub cache_js :Local Does(PseudoCache) PCTrueCache(1) {
       my ($self, $c) = @_;
       # Long running action to be cached
     }

     # used with Catalyst::Plugin::Cache and the optional key attr
     sub cache_with_key :Local Does(PseudoCache) PCTrueCache(1) PCKey('rememberme'){
       my ($self, $c) = @_;
       # Long running action to be cached
     }

DESCRIPTION
    This module was originally made to take the output of
    Catalyst::View::JavaScript::Minifier::XS and store it in a file so that
    after the server booted we would not need to generate it again and could
    let the static web server serve up the static file. Obviously it can be
    used for much more than javascript, but it's mostly made with large,
    purely javascript sites in mind. It does not cache the output of the
    action when the server is run in development mode.

ATTRIBUTES
  PCTrueCache
    Setting PCTrueCache will use Catalyst::Plugin::Cache and allow a real
    cache backend to do the work. After version 2 of this module this will
    no longer need to be set and the old mode of this plugin will be removed
    entirely.

  PCKey
    PCKey is an optional way of providing a different key for the cache
    backend. The default key is "Controller::Name/action".

    The two attributes below are DEPRECATED and provided for back compat
    only. They might disappear in the future. Using "PCTrueCache" and
    Catalyst::Plugin::Cache is highly recommended.

  PCUrl
    Required when not using "PCTrueCache".

    After the action runs once it will redirect to $PCUrl.

  PCPath
    When the action gets run the first time it will write it's output to
    $PCPath.

    Defaults to "$c->path_to('root') . $PCUrl"

    So using the example given above for the "all_js" action, the path will
    be

     $MyAppLocation/root/static/js/all.js

THANKS
    Thanks to Geoffrey Darling for writing all the code for the modern true
    cache mode

SEE ALSO
    Catalyst::Plugin::Cache

AUTHOR
    Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Arthur Axel "fREW" Schmidt.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Catalyst-ActionRole-PseudoCache

1.000002  2012-11-17 17:20:07 CST6CDT
  - Put MetaYAML back in dist

1.000001  2011-08-22 19:41:33 CST6CDT
  - Allow (and prefer) the use of Catalyst::Plugin::Cache (Geoff Darling)

0.093200 Mon Nov 16 21:22:00 CDT 2009
       - Don't delete tmp files in win32, it doesn't work

0.093050 Sun Nov 01 01:21:00 CDT 2009
       - Clarify POD

0.093030 Thu Oct 29 23:32:00 CDT 2009
       - Initial Release



t/lib/TestApp.pm
package TestApp;
our $VERSION = '0.093200';



use strict;
use warnings;

use Catalyst::Runtime 5.80;

use parent qw/Catalyst/;
use Catalyst qw/ Static::Simple /;

__PACKAGE__->config(
   name => 'TestApp',
);

__PACKAGE__->setup();

1;


t/lib/TestApp/Controller/Root.pm
package TestApp::Controller::Root;
our $VERSION = '0.093200';



use Moose;
BEGIN { extends 'Catalyst::Controller::ActionRole' };

__PACKAGE__->config->{namespace} = '';

has run_once => (
   is      => 'rw',
   isa     => 'Bool',
   default => undef,
);

sub test :Local :Does(PseudoCache) PCUrl(/foo.txt) {
    my ( $self, $c ) = @_;

    # this is so that we can be sure the output is cached
    unless ($self->run_once) {
       $c->response->body('big fat output');
    } else {
       $c->response->body('something else');
    }

    $self->run_once(1);
}

sub test2 :Local :Does(PseudoCache) PCUrl(/static/foo.txt) PCPath(bar.txt) {
    my ( $self, $c ) = @_;

    $c->response->body('something else');
}

sub end : Private :ActionClass(RenderView) {}

1;


Build.PL

use strict;
use warnings;

use Module::Build;

my $build = Module::Build->new(
  module_name   => 'Catalyst::ActionRole::PseudoCache',
  license       => 'perl',
  dist_abstract => "Super\ simple\ caching\ for\ Catalyst\ actions",
  dist_name     => "Catalyst\-ActionRole\-PseudoCache",
  dist_author   => [
"Arthur\ Axel\ \"fREW\"\ Schmidt\ \<frioux\+cpan\@gmail\.com\>",
  ],
  requires      => {
    "parent" => '0',
    "Catalyst::Plugin::Static::Simple" => '0',
    "autodie" => '0',
    "File::Spec" => '0',
    "Catalyst::Runtime" => '5.80013',
    "Catalyst::Controller::ActionRole" => '0.12',
  },
  build_requires      => {
    "Test::More" => '0.88',
    "FindBin" => '0',
    "Test::WWW::Mechanize::Catalyst" => '0',
    "Catalyst::Test" => '0',
  },
  script_files => [ qw() ],
);

$build->create_build_script;


lib/Catalyst/ActionRole/PseudoCache.pm
package Catalyst::ActionRole::PseudoCache;
{
  $Catalyst::ActionRole::PseudoCache::VERSION = '1.000002';
}

# ABSTRACT: Super simple caching for Catalyst actions

use Moose::Role;
use autodie;
use File::Spec;
use Carp qw(carp croak);

has true_cache => (
   is      => 'rw',
   isa     => 'Bool',
   default => undef,
);

has key => (
   is      => 'ro',
   isa     => 'Str',
   builder => '_build_key',
   lazy    => 1,
);

has is_cached => (
   is      => 'rw',
   isa     => 'Bool',
   default => undef,
);

has path => (
   is      => 'ro',
   isa     => 'Str',
   builder => '_build_path',
   lazy    => 1,
);

has url => (
   is       => 'ro',
   isa      => 'Str',
   required => 0,
);

sub _build_key {
   my $self = shift;
   return $self->class . '/' . $self->name;
}

sub _build_path {
   my $self = shift;
   my $url = $self->url;
   return File::Spec->catfile(split qr{/}, $url);
}

around BUILDARGS => sub {
   my $orig  = shift;
   my $class = shift;
   my ($args) = @_;
   if (my $attr = $args->{attributes}) {
      my @args;
      if ($attr->{PCTrueCache}) {
         @args = (
            true_cache => 1,
            ($attr->{PCKey}
               ? ( key => $attr->{PCKey}->[0] )
               : ()
            ),
            %{$args}
         );
         croak 'you must not set attributes that are not supported (PCUrl or PCPath) in true cache mode!'
            if $attr->{PCUrl} || $attr->{PCPath};
      } else {
         @args = (
            ($attr->{PCUrl}
               ? ( url => $attr->{PCUrl}->[0] )
               : ()
            ),
            ($attr->{PCPath}
               ? ( path => $attr->{PCPath}->[0] )
               : ()
            ),
            %{$args}
         );
         croak q(if you don't use true cache mode you must set PCUrl!)
            unless $attr->{PCUrl};

         carp 'you are not using true cache mode, which is pretty sucky, and more importantly, deprecated'
      }

      return $class->$orig( @args );
   } else {
      return $class->$orig(@_);
   }
};

around execute => sub {
   my $orig               = shift;
   my $self               = shift;
   my ( $controller, $c ) = @_;

   # do nothing if debug
   return $self->$orig(@_)
      if ($c->debug);

   if ($self->true_cache) {
      $self->_true_cache($orig,@_);
   } else {
       # backup method (for back compat)
      $self->_pseudo_cache($orig,@_);
   }
};

sub _true_cache {
   my ($self, $orig, $controller, $c, @rest) = @_;

   my $cache = $c->cache;

   my $body;
   unless ($body = $cache->get($self->key)){
      $self->$orig($controller, $c, @rest);
      $cache->set($self->key, $c->response->body);
   }
   $c->response->body($body);
}

sub _pseudo_cache {
   my ($self, $orig, $controller, $c, @rest) = @_;

   if (!$self->is_cached) {
      my $filename = File::Spec->catfile($c->path_to('root'), $self->path);

      unlink $filename if stat $filename;
      open my $js_fh, '>', $filename;

      $self->$orig($controller, $c, @rest);

      print {$js_fh} $c->response->body;
      close $js_fh;

      $self->is_cached(1);
   } else {
      $c->response->redirect($self->url, 300);
   }
}

1;

__END__

=pod

=head1 NAME

Catalyst::ActionRole::PseudoCache - Super simple caching for Catalyst actions

=head1 VERSION

version 1.000002

=head1 SYNOPSIS

 package MyApp::Controller::Root;

 use Moose;
 BEGIN { extends 'Catalyst::Controller::ActionRole' };

 # used with Catalyst::Plugin::Cache
 sub cache_js :Local Does(PseudoCache) PCTrueCache(1) {
   my ($self, $c) = @_;
   # Long running action to be cached
 }

 # used with Catalyst::Plugin::Cache and the optional key attr
 sub cache_with_key :Local Does(PseudoCache) PCTrueCache(1) PCKey('rememberme'){
   my ($self, $c) = @_;
   # Long running action to be cached
 }

=head1 DESCRIPTION

This module was originally made to take the output of
L<Catalyst::View::JavaScript::Minifier::XS> and store it in a file so that after
the server booted we would not need to generate it again and could let the
static web server serve up the static file.  Obviously it can be
used for much more than javascript, but it's mostly made with large, purely
javascript sites in mind.  It does not cache the output of the action when the
server is run in development mode.

=head1 ATTRIBUTES

=head2 PCTrueCache

Setting PCTrueCache will use L<Catalyst::Plugin::Cache> and allow a real
cache backend to do the work.  After version 2 of this module this will no
longer need to be set and the old mode of this plugin will be removed entirely.

=head2 PCKey

PCKey is an optional way of providing a different key for the cache backend.
The default key is C<Controller::Name/action>.

The two attributes below are B<DEPRECATED> and provided for back compat only. They
might disappear in the future. Using C<PCTrueCache> and L<Catalyst::Plugin::Cache>
is highly recommended.

=head2 PCUrl

Required when not using L</PCTrueCache>.

After the action runs once it will redirect to C<$PCUrl>.

=head2 PCPath

When the action gets run the first time it will write it's output to C<$PCPath>.

Defaults to C<< $c->path_to('root') . $PCUrl >>

So using the example given above for the C<all_js> action, the path will be

 $MyAppLocation/root/static/js/all.js

=head1 THANKS

Thanks to Geoffrey Darling for writing all the code for the modern true cache
mode

=head1 SEE ALSO

L<Catalyst::Plugin::Cache>

=head1 AUTHOR

Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Arthur Axel "fREW" Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Super simple caching for Catalyst actions",
  "AUTHOR" => "Arthur Axel \"fREW\" Schmidt <frioux+cpan\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Catalyst-ActionRole-PseudoCache",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Catalyst::ActionRole::PseudoCache",
  "PREREQ_PM" => {
    "Catalyst::Controller::ActionRole" => "0.12",
    "Catalyst::Plugin::Cache" => 0,
    "Catalyst::Plugin::Static::Simple" => 0,
    "Catalyst::Runtime" => "5.80013",
    "Catalyst::Test" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::More" => "0.88",
    "Test::WWW::Mechanize::Catalyst" => 0,
    "autodie" => 0,
    "parent" => 0
  },
  "VERSION" => "1.000002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



